crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -0,0 +1,561 @@
+                            uniform_sell_price,
+                            uniform_buy_price,
+                            *max_volume_factor,
+                        )?;
+                        // take the smaller of the two
+                        Some(std::cmp::min(price_improvement_fee, max_volume_fee))
+                    }
+                    (
+                        FeePolicy::PriceImprovement {
+                            factor,
+                            max_volume_factor,
+                        },
+                        Side::Sell,
+                    ) => {
+                        let price_improvement_fee = price_improvement_fee_for_sell(
+                            order.sell.amount.0,
+                            order.buy.amount.0,
+                            executed.0,
+                            surplus_fee,
+                            uniform_sell_price,
+                            uniform_buy_price,
+                            *factor,
+                        )?;
+                        let max_volume_fee =
+                            volume_fee_for_sell(executed.0, surplus_fee, *max_volume_factor)?;
+                        // take the smaller of the two
+                        Some(std::cmp::min(price_improvement_fee, max_volume_fee))
+                    }
+                    (FeePolicy::Volume { factor }, Side::Buy) => volume_fee_for_buy(
+                        executed.0,
+                        surplus_fee,
+                        uniform_sell_price,
+                        uniform_buy_price,
+                        *factor,
+                    ),
+                    (FeePolicy::Volume { factor }, Side::Sell) => {
+                        volume_fee_for_sell(executed.0, surplus_fee, *factor)
+                    }
+                }
+            })
+            .fold(eth::U256::zero(), |acc, fee| match fee {
+                Some(fee) => acc + fee,
+                None => {
+                    tracing::warn!(?order.uid, "failed to calculate protocol fee");
+                    acc
+                }
+            });
+
+        // Increase the fee by the protocol fee
+        let fee = match self.raw_fee() {
+            Fee::Static => Fee::Static,
+            Fee::Dynamic(fee) => Fee::Dynamic((fee.0 + protocol_fee).into()),
+        };
+
+        // Adjust the executed amount by the protocol fee. This is because solvers are
+        // unaware of the protocol fee that driver introduces and they only account
+        // for the surplus fee.
+        let executed = match order.side {
+            order::Side::Buy => executed,
+            order::Side::Sell => order::TargetAmount(
+                executed
+                    .0
+                    .checked_sub(protocol_fee)
+                    .ok_or(InvalidExecutedAmount)?,
+            ),
+        };
+
+        Fulfillment::new(order, executed, fee)
+    }
+}
+
+fn price_improvement_fee_for_buy(
```

Comment:
Can we make these methods operate on self (to avoid passing 5 extra arguments) and have one method for both buy/sell to further simplify the remaining very large match statement in the method above?

Reply:
Refactored.